### PR TITLE
Add composer template

### PR DESCRIPTION
**Pull request by xtreamwayz**
As discussed in several communication channels, there is no clear path on how the composer.json file should look like these days. There have been many changes which were not communicated to all maintainers and contributors in a central place. This PR adds a composer.json template to provide that central communication to everyone involved.

### CODE REVIEW

**Comment by froschdesign** [comment]
@xtreamwayz 
You are too fast! 😉 
I had already written a new issue report, but your PR is much better!

File: template/composer.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        "forum": "https://discourse.zendframework.com"
+    },
+    "extra": {
+        "branch-alias": {
+            "dev-master": "1.0-dev",
+            "dev-develop": "2.0-dev"
+        }
+    },
+    "require": {
+        "php": "^5.6 || ^7.0"
```
Comment (webimpress) [inline template/composer.json]: Should we change order also here, to have first `7.0`?

File: template/composer.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
+{
+    "name": "zendframework/<project-name>",
+    "description": "<project-description>",
+    "homepage": "https://docs.zendframework.com/<project-name>/",
+    "type": "library",
+    "license": "BSD-3-Clause",
+    "keywords": [
+        "http",
```
Comment (froschdesign) [inline template/composer.json]: I would replace these lines with a short description like.

```json
"keyword" : [
    "<individual keywords for current project>"
],
```

File: template/composer.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        "psr",
+        "psr-7",
+        "psr-11"
+    ],
+    "support": {
+        "docs": "https://docs.zendframework.com/<project-name>/",
+        "issues": "https://github.com/zendframework/<project-name>/issues",
+        "source": "https://github.com/zendframework/<project-name>",
+        "slack": "https://zendframework-slack.herokuapp.com",
+        "forum": "https://discourse.zendframework.com"
```
Comment (webimpress) [inline template/composer.json]: Should we have here link to the main page of forum or maybe specific category there?
https://github.com/zendframework/zend-expressive/blob/master/composer.json#L19

File: template/composer.json (hunk 2 above)
Comment (webimpress) [inline template/composer.json]: Maybe we should have here also `zf` or `zendframework` keyword?

File: template/composer.json [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            "@cs-check",
+            "@test"
+        ],
+        "upload-coverage": "coveralls -v",
+        "cs-check": "phpcs",
+        "cs-fix": "phpcbf",
+        "test": "phpunit --colors=always",
+        "test-coverage": "phpunit --coverage-clover clover.xml"
+    }
+}
```
Comment (webimpress) [inline template/composer.json]: Please add empty line at the end of the file.

File: template/composer.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        "branch-alias": {
+            "dev-master": "1.0-dev",
+            "dev-develop": "2.0-dev"
+        }
+    },
+    "require": {
+        "php": "^5.6 || ^7.0"
+    },
+    "require-dev": {
+        "phpunit/phpunit": "^6.0.8 || ^5.7.15",
```
Comment (webimpress) [inline template/composer.json]: I would update these version to the latest: `^6.2.3 || ^5.7.21`

File: template/composer.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
+{
+    "name": "zendframework/<project-name>",
+    "description": "<project-description>",
+    "homepage": "https://docs.zendframework.com/<project-name>/",
+    "type": "library",
+    "license": "BSD-3-Clause",
+    "keywords": [
+        "<individual keywords for current project>"
```
Comment (webimpress) [inline template/composer.json]: Maybe we should have here some common keywords like `zf` or `zendframework`?

File: template/composer.json [hunk 7]
```diff
@@ -0,0 +1,55 @@
+{
+    "name": "zendframework/<project-name>",
+    "description": "<project-description>",
```
Comment (froschdesign) [inline template/composer.json]: My suggestion:

```json
"description": "<project description, the same like the repository description>",
```

File: template/composer.json (hunk 6 above)
Comment (froschdesign) [inline template/composer.json]: Good question. The keywords are used at searching and filtering. Maybe is "zend" as mark / brand also interesting?!

File: template/composer.json (hunk 3 above)
Comment (froschdesign) [inline template/composer.json]: If a category exists, then we should use it.

File: template/composer.json (hunk 5 above)
Comment (xtreamwayz) [inline template/composer.json]: I agree, however PHP Unit updates twice a month or so. It should not be the goal of this template to keep track of PHP Unit releases.

File: template/composer.json (hunk 1 above)
Comment (xtreamwayz) [inline template/composer.json]: Don't know. I think in every project it's the lowest PHP first at the moment.

File: template/composer.json [hunk 8]
```diff
@@ -2,7 +2,6 @@
     "name": "zendframework/<project-name>",
     "description": "<project description, the same like the repository description>",
     "homepage": "https://docs.zendframework.com/<project-name>/",
-    "type": "library",
```
Comment (froschdesign) [inline template/composer.json]: Good catch! 👍

**Comment by webimpress** [comment]
@xtreamwayz what about adding:
```json

"config": {
  "sort-packages": true
},
```
https://github.com/weierophinney/hal/blob/master/composer.json#L15-L17

**Comment by xtreamwayz** [comment]
I was thinking about `sort-packages` as well. I'm not sure how that works out with existing dependencies. It's added to the expressive skeleton as well. It does make it easier to lookup dependencies when they are sorted.

**Comment by xtreamwayz** [comment]
Currently version constrains are written in different ways:

- `"^5.6 || ^7.0"` 
- `"^6.2 || ^5.7",`
- `"^1.0 || 1.0.x-dev"`.

According to the composer docs it doesn't matter how you write it?

> When you write a version constraint, it may reference a specific tag (e.g., 1.1) or it may reference a valid range of tags (e.g., >=1.1 <2.0, or ~4.0). To resolve these constraints, Composer first asks the VCS to list all available tags, then creates an internal list of available versions based on these tags. In the above example, composer's internal list includes versions 1.0, 1.0.1, 1.0.2, the beta release of 1.1, the first and second release candidates of 1.1, the final release version 1.1, etc....
>
> When Composer has a complete list of available versions from your VCS, it then finds the highest version that matches all version constraints in your project.

So what to do? `"lowest || highest"` vs `"highest || lowest"`.

**Comment by froschdesign** [comment]
> According to the composer docs it doesn't matter how you write it?

Right, because: 

> …finds the highest version…

I prefer a simple and intuitive reading: `"lowest || highest"`

**Comment by weierophinney** [comment]
I assumed that the `||` operator worked as it does in most programming languages, choosing the first to match. However, when I consider it from a practical point of view, it will use this to determine what _set_ of packages are valid, so it can choose the latest based on the requirements of other dependencies. 

So, the main thing is that we are consistent. I agree with @froschdesign that latest last makes sense, as the constraints are then ordered.

**Comment by froschdesign** [comment]
LGTM 👍

File: template/composer.json [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        "zend",
+        "zendframework",
+        "<package_name>",
+        "<individual keywords for current package>"
+    ],
+    "support": {
+        "docs": "https://docs.zendframework.com/<package_name>/",
+        "issues": "https://github.com/zendframework/<package_name>/issues",
+        "source": "https://github.com/zendframework/<package_name>",
+        "rss": "https://github.com/zendframework/<package_name>/releases.atom",
```
Comment (weierophinney) [inline template/composer.json]: I find it funny that we use an Atom feed for the RSS item. :smile: That said, it's the only appropriate location.

File: template/composer.json [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        "slack": "https://zendframework-slack.herokuapp.com",
+        "forum": "https://discourse.zendframework.com/c/questions/<components|expressive|apigility>"
+    },
+    "require": {
+        "php": "^5.6 || ^7.0"
```
Comment (weierophinney) [inline template/composer.json]: For new packages, this will be simply `^7.1`, in which case, we can change the PHPUnit constraint to simply `^6.2`.

**Comment by weierophinney** [comment]
Thanks, @webimpress!

**Comment by xtreamwayz** [comment]
> Thanks, @webimpress!

???

**Comment by webimpress** [comment]
lol, thanks @xtreamwayz  :+1: :+1: :+1:

**Comment by weierophinney** [comment]
@xtreamwayz I typed the wrong issue identifier when I wrote that! (I use [gh](https://github.com/node-gh/gh) from the command line for a number of tasks)

THANK YOU, Geert!

**Pull request merged**